src/main/java/org/jenkinsci/plugins/workflow/support/storage/BulkFlowNodeStorage.java

```diff
@@ -0,0 +1,247 @@
+    private boolean isModified = false;
+
+    File getStoreFile() throws IOException {
+        return new File(dir, "flowNodeStore.xml");
+    }
+
+    public BulkFlowNodeStorage(FlowExecution exec, File dir) {
+        this.exec = exec;
+        this.dir = dir;
+        this.nodes = null;
+        this.setAvoidAtomicWrite(true);
+    }
+
+    /** Loads the nodes listing, lazily - so loading the {@link FlowExecution} doesn't trigger a more complex load. */
+    HashMap<String, Tag> getOrLoadNodes() throws IOException {
+        if (nodes == null) {
+            if (dir.exists()) {
+                File storeFile = getStoreFile();
+                if (storeFile.exists()) {
+                    HashMap<String, Tag> roughNodes = (HashMap<String, Tag>) (XSTREAM.fromXML(getStoreFile()));
+                    if (roughNodes == null) {
+                        throw new IOException("Unable to load nodes, invalid data");
+                    }
+                    for (Tag t : roughNodes.values()) {
+                        FlowNode fn = t.node;
+                        try {
+                            FlowNode$exec.set(fn, exec);
+                        } catch (IllegalAccessException e) {
+                            throw (IllegalAccessError) new IllegalAccessError("Failed to set owner").initCause(e);
+                        }
+                        t.storeActions();
+                        for (FlowNodeAction a : Util.filter(t.actions(), FlowNodeAction.class)) {
+                            a.onLoad(fn);
+                        }
+                    }
+                    nodes = roughNodes;
+                } else {
+                    nodes = new HashMap<String, Tag>();
+                }
+            } else {
+                IOUtils.mkdirs(dir);
+                nodes = new HashMap<String, Tag>();
+            }
+        }
+        return nodes;
+    }
+
+    @Override
+    @CheckForNull
+    public FlowNode getNode(@Nonnull String id) throws IOException {
+        Tag t = getOrLoadNodes().get(id);
+        return (t != null) ? t.node : null;
+    }
+
+    public void storeNode(@Nonnull FlowNode n, boolean delayWritingActions) throws IOException {
+        Tag t = getOrLoadNodes().get(n.getId());
+        if (t != null) {
+            t.node = n;
+            List<Action> act = n.getActions();
+            t.actions = act.toArray(new Action[act.size()]);
+        } else {
+            getOrLoadNodes().put(n.getId(), new Tag(n, n.getActions()));
+        }
+        isModified = true;
+        if (!delayWritingActions) {
+            flush();
+        }
+    }
+
+    @Override
+    public void storeNode(FlowNode n) throws IOException {
+        storeNode(n, false);
+    }
+
+    /**
+     * Persists a single FlowNode to disk (if not already persisted).
+     * @param n Node to persist
+     * @throws IOException
+     */
+    @Override
+    public void flushNode(@Nonnull FlowNode n) throws IOException {
+        storeNode(n, false);
```

Comment:
My gut instinct on identical methods like this and `storeNode(FlowNode)` above is to have one call the other. No big deal, I just find it prevents later confusion.

Reply:
done, good point!